### DISCUSSION TITLE
It’s time to end the hate (for bool)

### DISCUSSION THREAD

**Discussion by JimB123** [valkey-io/valkey#1278] [General]
**Can anyone explain why we continue to persecute the lowly `bool`?**

The boolean type was formally added in C99 (25 years ago!).  Yes, it required a special header, but so do a lot of things.  As of C23, bool is a first class type, using reserved words for `bool`/`true`/`false`.  Yet bool continues to be discriminated against in Valkey.

Booleans are more readable and expressive.  When there’s an `int` representing 0/1/false/true - a reader never really knows for sure that it’s ALWAYS 0/1 (or even that 0/1 necessarily represent false/true).  A boolean conveys a clear intent - AND THE COMPILER ENFORCES IT.

I’ve heard people say “We don’t use bool because \<insert name\> doesn’t like bool”.  But should we give into peer pressure and join the haters?  I say NO!

This is Valkey now.  It’s a brave new world.  Let’s shed the shackles of “bool hate”.  Rise up and be a force for love in the world!

Let’s face it, we use `goto` all over the place - and that’s the evil construct that our computer science instructors all warned us about.  We use other C99 additions like `inline` functions and `long long`.  We use `enum`.  Is it somehow more palatable to do this?
```
typedef enum {notABool_False=0, notABool_True=1} NotABool;  // Just silliness
```

As one of the few people here who was actually using C before C99, I say enough is enough.

_‘Cause the good old days weren’t always good.
And tomorrow ain’t as bad as it seems._
     - Billy Joel

**Comment by zuiderkwast** [comment]
The only logical explanation I've heard is that if you're blindly replacing `int` with `bool`, it can have some side effect because `sizeof(bool) != sizeof(int)`. If we're just aware of that, then it won't be a problem IMHO.

@valkey-io/core-team I marked this as a major decision. Let's vote about it. ( :+1: / :-1: )

**Reply by enjoy-binbin** [reply]
i don't mind doing this, the only painful thing for me just the git blame and the git conflict, but we already done a lot of formatting and refactoring things.

**Reply by JimB123** [reply]
Note that I'm **NOT** recommending replacing `int` with `bool`.  I'm just recommending that we stop saying "hey, don't use bool" in our code reviews.  Designs shouldn't be limited to using `int` when a `bool` is the logically correct.

I'm also not recommending blindly casting between int/bool.

Just start using `bool` when it's appropriate to do so.

**Reply by hwware** [reply]
I do some research on it (Because I have no experience using bool in C), I think we can use it in Valkey, but we need to do review more carefully on it.  Ref: https://softwareengineering.stackexchange.com/questions/291771/what-was-the-reason-for-the-creation-of-boolean-variables and https://news.ycombinator.com/item?id=6309008

**Reply by hwware** [reply]
> Note that I'm **NOT** recommending replacing `int` with `bool`. I'm just recommending that we stop saying "hey, don't use bool" in our code reviews. Designs shouldn't be limited to using `int` when a `bool` is the logically correct.
> 
> I'm also not recommending blindly casting between int/bool.
> 
> Just start using `bool` when it's appropriate to do so.

yes, we can not just simply replace int with bool, we need to care about the logic first.

**Comment by zuiderkwast** [comment]
The majority of TSC have voted for it, so consider this accepted.
